reconcile/utils/runtime/desired_state_diff.py

```diff
@@ -0,0 +1,161 @@
+import logging
+import multiprocessing
+from dataclasses import dataclass
+from typing import (
+    Any,
+    Callable,
+    Optional,
+)
+
+from deepdiff import DeepHash
+from jsonpath_ng.ext.parser import parse
+
+from reconcile.change_owners.diff import (
+    Diff,
+    DiffType,
+    extract_diffs,
+)
+from reconcile.utils.jsonpath import apply_constraint_to_path
+from reconcile.utils.runtime.integration import (
+    DesiredStateShardConfig,
+    ShardedRunProposal,
+)
+
+
+@dataclass
+class DesiredStateDiff:
+
+    previous_desired_state: Any
+    current_desired_state: Any
+    diff_found: bool
+    affected_shards: set[str]
+
+    def can_exit_early(self) -> bool:
+        return not self.diff_found
+
+    def can_run_sharded(self) -> bool:
+        return self.affected_shards != set()
+
+
+def find_changed_shards(
+    diffs: list[Diff],
+    previous_desired_state: dict[str, Any],
```

Comment:
Reminder on abstracts over concrete inputs (concrete outputs are fine).

`dict` -> `Mapping`
`list` -> `Iterable`

and all that good stuff unless there's a good reason not to.